JSDK-2408 add room and participant sid in connect message.

makarandp0 opened the pull request:
@manjeshbhargav, @innerverse 

This change waits for roomSid, partcipantSid before connecting to event publisher endpoint. That way we can pass roomSid/participantSid along with the connect message. I have few open question around this (https://twilio.slack.com/archives/C35JTJHJP/p1563392120206400), but I would like to get initial feedback. 

TODO:
- [x] add unit tests
 
- [x] I acknowledge that all my contributions will be made under the project's license.

manjeshbhargav commented on lib/signaling/v2/twilioconnectiontransport.js:
```diff
@@ -449,6 +453,8 @@ function setupEventListeners(transport) {
             transport._session = message.session;
             transport.emit('connected', message);
+            // update eventPublisher with participant sid, and room
+            // transport.updateEventPublisher(message.sid, message.participant.sid);
```
@makarandp0 Can you remove these comments before merging?

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -46,6 +47,9 @@ class InsightsPublisher extends EventEmitter {
       _eventQueue: {
         value: []
       },
+      _readyToPublish: {
```
@makarandp0 Nitpick: `_readyToConnect`?

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -71,7 +75,18 @@ class InsightsPublisher extends EventEmitter {
+   * @param {string} roomSid
+   * @param {string} participantSid
+   */
+  update(roomSid, participantSid) {
```
@makarandp0 Nitpick:
```js
class InsightsPublisher {
  /**
   * Start connecting to the Insights gateway.
   * @param {string} roomSid
   * @param {string} participantSid
   * @returns {void}
   */
  connect(roomSid, participantSid) {
    ...
  }
}
```

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -71,7 +75,18 @@ class InsightsPublisher extends EventEmitter {
       self.removeListener('disconnected', maybeReconnect);
```
@makarandp0 Just above this line, can you also make sure `roomSid` and `participantSid` are passed to the `reconnect()` function as well?

makarandp0 replied:
thank you @manjeshbhargav for catching this!

manjeshbhargav commented on lib/signaling/v2/twilioconnectiontransport.js:
```diff
@@ -172,6 +172,10 @@ class TwilioConnectionTransport extends StateMachine {
     });
     setupEventListeners(this);
+
+    this.once('connected', (initialState) => {
```
@makarandp0 Nitpick:
```js
this.once('connected', ({ sid, participant }) => {
  this._eventPublisher.connect(sid, participant.sid);
});
```
?

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -59,19 +63,31 @@ class InsightsPublisher extends EventEmitter {
-      }
-      self.removeListener('disconnected', maybeReconnect);
+    this._readyToConnect.promise.then(({ roomSid, participantSid }) => {
+      const self = this;
```
@makarandp0 This can move out of the promise resolver.

makarandp0 replied:
not sure what you mean - we do need `maybeReconnect` inside the resolver because it references `roomSid, participantSid` which are available only after promise is resolved.

regarding `self` that is inside the resolver because outside code does not need it.

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -89,7 +105,10 @@ class InsightsPublisher extends EventEmitter {
    */
   disconnect() {
-    if (this._ws.readyState === this._WebSocket.CLOSING || this._ws.readyState === this._WebSocket.CLOSED) {
+    if (
```
@makarandp0 Nitpick: can you refactor so that the operators are at the beginning of the line:
```js
if (this._ws === null
  || this._ws.readyState === this._WebSocket.CLOSING
  || this._ws.readyState === this._WebSocket.CLOSED)
```
?

manjeshbhargav commented on lib/util/insightspublisher/index.js:
```diff
@@ -111,7 +130,10 @@ class InsightsPublisher extends EventEmitter {
    */
   publish(groupName, eventName, payload) {
-    if (this._ws.readyState === this._WebSocket.CLOSING || this._ws.readyState === this._WebSocket.CLOSED) {
+    if (
```
@makarandp0 Nitpick: can you refactor so that the operators are at the beginning of the line:
```js
if (this._ws !== null
  && (this._ws.readyState === this._WebSocket.CLOSING
    || this._ws.readyState === this._WebSocket.CLOSED))
```
?

makarandp0 commented on lib/util/insightspublisher/index.js:
```diff
@@ -59,19 +63,31 @@ class InsightsPublisher extends EventEmitter {
+        self.removeListener('disconnected', maybeReconnect);
+      });
+      connect(this, token, sdkName, sdkVersion, roomSid, participantSid, options);
+    }).catch(() => {
```
passing rejection handler as 2nd param to .then is a anti-pattern that should be avoided.

makarandp0 commented:
@manjeshbhargav, this is ready for another look. Please take a look when you get a chance

manjeshbhargav commented on test/integration/spec/util/insightspublisher.js:
```diff
@@ -31,13 +31,15 @@ describe('InsightsPublisher', function() {
 
       context(`when attempted with ${a(tokenType)} ${tokenType} token`, () => {
-        before(() => {
+        before( () => {
```
@makarandp0 Nitpick: remove extra space between the two `(`s?

manjeshbhargav commented on test/integration/spec/util/insightspublisher.js:
```diff
@@ -31,13 +31,15 @@ describe('InsightsPublisher', function() {
             options.environment,
             'us1',
             options);
+
```
@makarandp0 Nitpick: remove extra line?

manjeshbhargav commented on test/integration/spec/util/insightspublisher.js:
```diff
@@ -73,6 +75,8 @@ describe('InsightsPublisher', function() {
         options);
 
+      publisher.connect('roomSid', 'participantSid');
+
```
@makarandp0 Nitpick: remove extra line?

makarandp0 commented:
Thanks for catching the issue with `NullInsightsPublisher` @manjeshbhargav. I was able to repro it with an integration test, and its now fixed. I also added the test that could have caught the issue.
